farming/logics/traits/master_chef/farming.rs

```diff
@@ -0,0 +1,482 @@
+                            .ok_or(FarmingError::SubUnderflow4)?
+                            .checked_mul(self._arsw_per_block(period)?)
+                            .ok_or(FarmingError::MulOverflow3)?
+                            .checked_mul(pool_info.alloc_point as u128)
+                            .ok_or(FarmingError::MulOverflow4)?
+                            .into(),
+                    )
+                    .ok_or(FarmingError::AddOverflow6)?
+                    .checked_div(total_alloc_point.into())
+                    .ok_or(FarmingError::DivByZero1)?
+            } else {
+                arsw_reward = arsw_reward
+                    .checked_add(
+                        (self._period_max(period)? as u128)
+                            .checked_sub(last_block.into())
+                            .ok_or(FarmingError::SubUnderflow5)?
+                            .checked_mul(self._arsw_per_block(period)? as u128)
+                            .ok_or(FarmingError::MulOverflow5)?
+                            .checked_mul(
+                                pool_info
+                                    .alloc_point
+                                    .checked_div(total_alloc_point)
+                                    .ok_or(FarmingError::DivByZero2)?
+                                    .into(),
+                            )
+                            .ok_or(FarmingError::MulOverflow6)?
+                            .into(),
+                    )
+                    .ok_or(FarmingError::AddOverflow7)?;
+                last_block = self._period_max(period)?;
+            }
+
+            period += 1;
+        }
+
+        Ok(arsw_reward
+            .checked_mul(ACC_ARSW_PRECISION)
+            .ok_or(FarmingError::MulOverflow8)?
+            .checked_div(lp_supply)
+            .ok_or(FarmingError::DivByZero3)?)
+    }
+
+    fn _get_period(&self, block_number: u32) -> Result<u32, FarmingError> {
+        if block_number < ARTHSWAP_ORIGIN_BLOCK {
+            return Err(FarmingError::BlockNumberLowerThanOriginBlock)
+        }
+
+        // BLOCK_PER_PERIOD is never 0
+        return Ok(block_number
+            .checked_sub(ARTHSWAP_ORIGIN_BLOCK)
+            .ok_or(FarmingError::SubUnderflow1)?
+            / BLOCK_PER_PERIOD)
+    }
+
+    fn _period_max(&self, period: u32) -> Result<u32, FarmingError> {
+        Ok(ARTHSWAP_ORIGIN_BLOCK
+            .checked_add(
+                BLOCK_PER_PERIOD
+                    .checked_mul(period.checked_add(1).ok_or(FarmingError::AddOverflow4)?)
+                    .ok_or(FarmingError::MulOverflow1)?,
+            )
+            .ok_or(FarmingError::AddOverflow5)?
+            .checked_sub(1)
+            .ok_or(FarmingError::SubUnderflow3)?)
+    }
+
+    fn _arsw_per_block(&self, period: u32) -> Result<Balance, FarmingError> {
+        if period > MAX_PERIOD {
+            return Ok(0)
+        }
+        Ok(FIRST_PERIOD_REWERD_SUPPLY
+            .checked_mul(
```

Comment:
Better use casted_mul to ensure precision.

Reply:
But we need to cast back to u128 so casted_mul will not give more precision here

Reply:
Since we divide it with 10**period, it will be fine; for multiplication, use casted_mul and cast back after division.